<script src="feature-viewer.nextprot.js"></script>
<!-- Latest compiled and minified CSS -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

<!-- Optional theme -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

<!-- Latest compiled and minified JavaScript -->
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<style type="text/css">
    .svgHeader {
        display: none;
    }
    select {
        margin: 10px;
    }
    table {
        margin: 10px;
    }
    .geneContainer {
        margin: 10px;
        padding: 10px;
        border-width: 1px;
        border-color: grey;
        border-style: solid;
    }
    table tr td {
        border-width: 1px;
        border-color: grey;
        border-style: solid;
    }
</style>

<body>
      <div style="margin: 10px; font-size:12px">
        <h3>Normalisation parameters (play with it to find a nice fit for the game)</h3>
        <p>Introns are the non-coding parts of the gene (represented as the holes), which we could classify in 5 different classes for the game:</p>
        <ul>
            <li><label>small-intron is smaller than </label><input id="small-intron" type="number" value="5000">&nbsp;bps and normalized value is:<input id="small-intron-norm" type="number" value="1">units. With this type of intron we could say the avatar could easily jump.</li>
            <li><label>medium-intron is smaller than </label><input id="medium-intron" type="number" value="10000">&nbsp;bps and normalized value is:<input id="medium-intron-norm" type="number" value="5">units. With this type of intron we could say there are 1, 2 splicies (rocks) that the avatar need to jump on top to traverse the holw</li>
            <li><label>large-intron is smaller than </label><input id="large-intron" type="number" value="15000">&nbsp;bps and normalized value is:<input id="large-intron-norm" type="number" value="10">units. With this type of intron we could say that there is a spliceosome that moves</li>
            <li><label>big-intron is smaller than </label><input id="big-intron" type="number" value="20000">&nbsp;bps and normalized value is:<input id="big-intron-norm" type="number" value="20">units.  With this type of intron we could say that there is a spliceosome that moves and it is more complicated</li>
            <li><label>huge-intron is equal or bigger than </label><input id="huge-intron" type="number" value="20000">&nbsp;bps and normalized value is:<input id="huge-intron-norm" type="number"  value="50">units. With this type of intron there could be an SR protein before and the avatar would be propulsated (I think it is easy for the avatar)</li>
        </ul>
        <p>Exons (includes stop_codon, snp, start_codon) are relatively smallers (average is 500bps) compared to the introns (10000bps) and a continous normalisation approach was adopted:</p>
        <ul>
            <li><label>Division factor </label><input id="division-factor-exon" type="number" value="20">&nbsp; The length of the exons will be divided by this factor</li>
            <li><label>Minimal unit: </label><input style="font-size: 16px" id="minimal-unit" type="number" value="1">&nbsp; The exons will have in minimum this unit</li>
            <li><label>Maximal unit: </label><input style="font-size: 16px" id="maximal-unit" type="number" value="100">&nbsp; The exons will have in maximum this unit</li>
        </ul>
        <button onclick="loadProtein()" style="margin: 5px; font-size: 15px">Click here to normalize</button>
        <br>
    </div>
    <div style="margin: 10px;">
        Based on your parameters: 
        Minimal units for a gene are: <span id="minUnits" style="color:green">computing...</span>
        Maximal units for a gene are: <span id="maxUnits" style="color:green">computing...</span>
        <br>Exons count with minimal unit: <span id="minExonsCount" style="color:green">computing...</span>
        Exons count with maximal unit: <span id="maxExonsCount" style="color:green">computing...</span> (minimize this number)
        <br>Introns types distribution: <span id="intronsDist" style="color:green">computing...</span>
        
    </div>
    <div style="margin: 10px;">
        <h3>Notes</h3>
        <ul>
            <li>Up / Down / Obstacles are not yet defined (not sure if we should provide this info or we leave it up to you)....</li>
            <li>Intron are much bigger than exons in reality (if possible in the drawing of the blue lines could be a brin "enroule" to seem bigger...</li>
            <li>There is not a hole below the start, snp stop codon (it is a limitation of the library in use). It should be one exon with different color: from exon to utr (untranslated region)</li>
        </ul>
    </div>
    <div id="container">
      <div id="genes"></div>
    </div>
</body>


<script>

    var maximalExon, minimalExon = 0;

    loadProtein()

    function getBaseLog(x, y) {
        return Math.log(y) / Math.log(x);
    }
    
    function normalizeLength(levelData){
        
        if (levelData.type === "intron") {
            var normalizedIntron = getNormalizedLengthForIntrons(levelData.realLength);
            levelData.class = normalizedIntron.class
            levelData.normalizedLength = normalizedIntron.normalizedLength
        } else {
            var normalized = getNormalizedLengthForNonIntron(levelData.realLength);
            levelData.normalizedLength = normalized.normalizedLength
            levelData.class = levelData.type
        }
        
    }

    function getNormalizedLengthForNonIntron(realLength) {

        var divisionFactor = parseInt($("#division-factor-exon").val())
        var minimalUnit = parseInt($("#minimal-unit").val())
        var maximalUnit = parseInt($("#maximal-unit").val())

        //Based on histogram in R
        //hist(log(exons$V6), breaks=5, col="lightgreen")
        var continousNormalization = Math.round(realLength / divisionFactor);
        if (continousNormalization > maximalUnit) {
            continousNormalization = maximalUnit;
            maximalExon++;
        }
        if (continousNormalization < minimalUnit) {
            continousNormalization = minimalUnit;
            minimalExon++;
        }
        

        return { normalizedLength: continousNormalization }

    };

    function getProteinLink (gene, source){

                if(gene.externalLinks && gene.externalLinks[source]){
                    var externalLink = gene.externalLinks[source];
                   return  "<li><a target='_blank' href='"+ externalLink["url"] + "'>"+
                                 "<img height='20px' src='" + externalLink["img"] + "'</img>"+
                                 "</a> "+ externalLink["accession"]  + " </li>";
                }else return "";

    }


    function getNormalizedLengthForIntrons(number) {

        //Based on histogram in R
        //hist(log(exons$V6), breaks=5, col="lightgreen")
        if (number < parseInt($("#small-intron").val()))
            return {
                class: "small-intron",
                normalizedLength: parseInt($("#small-intron-norm").val())
            } //simple jump --> most of them end up in this category
        if (number < parseInt($("#medium-intron").val()))
            return {
                class: "medium-intron",
                normalizedLength: parseInt($("#medium-intron-norm").val())
            } //mutiple jumps (kind of rocks) 
        if (number < parseInt($("#large-intron").val()))
            return {
                class: "large-intron",
                normalizedLength: parseInt($("#large-intron-norm").val())
            } //spliceosome
        if (number < parseInt($("#big-intron").val()))
            return {
                class: "big-intron",
                normalizedLength: parseInt($("#big-intron-norm").val())
            } //Complicated spliceosome
        return {
            class: "huge-intron",
            normalizedLength: parseInt($("#huge-intron-norm").val())
        } //SR protein (Catapult)
    };
    

    function renderGenes(genesInfo, genesData) {


        maximalExon = 0;
        minimalExon = 0;
        //Normalization and get max
        var maxNormalizedLength = 0;
        var minNormalizedLength = Number.MAX_SAFE_INTEGER;

        var totalClassCount = {};
        genesData.forEach(function (gene) {
            var classCount = {};
            gene.totalNormalizedLength = 0
            gene.totalNormalizedWalkableLength = 0
            gene.levelData.forEach(function (ld) {
                normalizeLength(ld);
                if(!classCount[ld.class]){
                    classCount[ld.class] = 0;
                } 
                if(!totalClassCount[ld.class]){
                    totalClassCount[ld.class] = 0;
                } 
                classCount[ld.class] += 1;
                totalClassCount[ld.class] += 1;
                gene.totalNormalizedLength += ld.normalizedLength
                if(ld.class.indexOf("intron") == -1 ){
                    gene.totalNormalizedWalkableLength += ld.normalizedLength
                }
            })

            gene.classCount = classCount;

            var gi = genesInfo.filter(g => g.geneName === gene.geneName);
            Object.assign(gene, gi[0])

            maxNormalizedLength = Math.max(maxNormalizedLength, gene.totalNormalizedLength)
            minNormalizedLength = Math.min(minNormalizedLength, gene.totalNormalizedLength)
        })
        
        genesSorted = genesData.sort(function(a,b){
            if (a.totalNormalizedLength < b.totalNormalizedLength) {
                return -1;
            }else if (b.totalNormalizedLength < a.totalNormalizedLength) {
                return 1;
            }
            return 0;
        });

        $("#maxUnits").text(maxNormalizedLength)
        $("#minUnits").text(minNormalizedLength)
        $("#maxExonsCount").text(maximalExon)
        $("#minExonsCount").text(minimalExon)
        $("#intronsDist").text("");
        $("#intronsDist").append("small-introns:" + totalClassCount["small-intron"] + " ")
        $("#intronsDist").append("medium-introns:" + totalClassCount["medium-intron"] + " ")
        $("#intronsDist").append("big-introns:" + totalClassCount["big-intron"] + " ")
        $("#intronsDist").append("large-introns:" + totalClassCount["large-intron"] + " ")
        $("#intronsDist").append("huge-introns:" + totalClassCount["huge-intron"] + " ")

        genesSorted.forEach(function (gene) {



            var sv = ""; 
            if(gene.superVariant) {
                sv = "<span style='background-color:orange'>super variant</span>";
            }

            var html = "";
            html += "<div class='geneContainer'>" +
                    "<div class='pull-right'><img width='50px' src='http://www.chromosomewalk.ch/wp-content/uploads/chromosome_" + gene.chromosome + ".png'></img></div>" +
                    "<h3>Gene name: " + gene.geneName  + "&nbsp;" + sv + "</h3>" + 
                    "<b>Transcript</b>: <a target='_blank' href='http://www.ensembl.org/Homo_sapiens/transview?db=core;transcript=" + gene.transcript + "'>" + gene.transcript + "</a>"+
                    "&nbsp;&nbsp;<b>Chromosome</b>: " + gene.chromosome + 
                    "<br><b>Gene real length:</b>" + gene.geneLength + " bp (base pairs = A,C,G,T)  " + 
                    "&nbsp;&nbsp;<b>Protein real length:</b> " + Math.ceil((gene.exonsLength + 1) / 3.0)  + "aa (amino acids). Computed using (exonsLength / 3) but seems not to be correct all the time" +
                    "<br><b>Normalized Length:</b>"+ gene.totalNormalizedLength + " units (walkable: " + gene.totalNormalizedWalkableLength + " units) composed by <b>exons:</b> " + gene.classCount.exon + 
                    "&nbsp;&nbsp;<b>small-introns:</b> " + (gene.classCount["small-intron"] ? gene.classCount["small-intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>medium-introns:</b> " + (gene.classCount["medium-intron"] ? gene.classCount["medium-intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>large-introns:</b> " + (gene.classCount["large-intron"] ? gene.classCount["large-intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>big-introns:</b> " + (gene.classCount["big-intron"] ? gene.classCount["big-intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>huge-introns:</b> " + (gene.classCount["huge-intron"] ? gene.classCount["huge-intron"] : "0")  + 
                    "<br>???% of protein coding (exons)" +
                    "&nbsp;Variant: rs000";

//           html += "<br><button data-toggle='collapse' data-target='#json" + gene.id + "'>View JSON in pretty format</button>";
//            html += "<div id='json" + gene.id + "' class='collapse'><pre><code>" + JSON.stringify(gene, null, 2) + "</code></pre></div>";
            html += "<br><span style='font-size:6px'>" + JSON.stringify(gene, null, 2) + "</span>";
            html += "<div id=\'" + gene.id + "\'></div>";
            html += "<b>Mutation Type</b>: (Missense, Insertion, Deletion, Frameshift, STOP, Intergenic, Others...)<br>";
            html += "<b>Description</b>:<br>" + gene.description ;
            if(gene["3dImage"])
                html += "<img width='250px' src='" + gene["3dImage"] + "'></img>";
            
            html += "<br><b>Protein Resources Databases:</b><br>" + 
                "<ul>" + getProteinLink(gene, "uniprot") + getProteinLink(gene, "nextprot") +
                "</ul>";
            html += "<br><b>Links to publication, chromosome walk, swissdrug?, ensembl? dbsnpedia? </b><br>"

            if(gene.debugInfo) {
                html += "<br><b>Debug Info Image (To help understanding but not to include in the game)</b><br>";
                html += "<img height='50px' src='" + gene.debugInfo.img + "'></img>";
            }

            html += "</div>";
            html += "<hr>";

            $("#genes").append(html);

            //Create a new Feature Viewer and add some rendering options
            var ft = new FeatureViewer(maxNormalizedLength, "#" + gene.id, {
                showAxis: true,
                showSequence: true,
                brushActive: true,
                toolbar: true,
                bubbleHelp: true,
                zoomMax: 10
            });

            var exonData = [];
            var intronData = [];
            var starsData = [];

            var cursor = 0;
            gene.levelData.forEach(function (ld) {
                
                var desc = ld.class ? ld.class : "" + "\n normalizedLength:" + ld.normalizedLength + "\n realLength:" + ld.realLength;
                var normalizedLength = ld.normalizedLength;
                if (ld.type === "exon" || ld.type === "utr") {
                    exonData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": (ld.type == "exon") ? "green" : "lightgreen"
                    })
                } else if (ld.type === "intron") {
                    intronData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": "blue"
                    })
                } else if (ld.type === "start_codon" || ld.type === "stop_codon" || ld.type === "snp") {
                    starsData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": (ld.type == "snp") ? "red" : "orange",
                        "description": ld.type
                    })

                }
                cursor += normalizedLength
            })

            //Add some features
            ft.addFeature({
                data: starsData,
                name: "stars",
                type: "rect"
            });

            //Add some features
            ft.addFeature({
                data: exonData,
                name: "exons",
                type: "rect"
            });

            //Add some features
            ft.addFeature({
                data: intronData,
                name: "introns",
                type: "rect"
            });

        })


    }

    function loadProtein() {
        $("#maxUnits").text("computing...")
        $("#minUnits").text("computing...")
        $("#intronsDist").text("computing...")
        $("#maxExonsCount").text("computing...")
        $("#minExonsCount").text("computing...")
        
        $("#genes").empty();
        $.getJSON('data/genes.json', function (genesInfo) {
            $.getJSON('data/data.json', function (genesData) {
                renderGenes(genesInfo, genesData);
            })
        });
    }
</script>