<script src="feature-viewer.nextprot.js"></script>
<!-- Latest compiled and minified CSS -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

<!-- Optional theme -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

<!-- Latest compiled and minified JavaScript -->
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<style type="text/css">
    .svgHeader {
        display: none;
    }
    select {
        margin: 10px;
    }
    table {
        margin: 10px;
    }
    .geneContainer {
        margin: 10px;
        padding: 10px;
        border-width: 1px;
        border-color: grey;
        border-style: solid;
    }
    table tr td {
        border-width: 1px;
        border-color: grey;
        border-style: solid;
    }
</style>

<body>
      <div style="margin: 10px; font-size:12px">
        <h3>Normalisation parameters (play with it to find a nice fit for the game)</h3>
        <br><button data-toggle='collapse' data-target="#normParams">Change normalization parameters</button>
        <div id="normParams" class='collapse'>
            <p>Introns are the non-coding parts of the gene (represented as the holes), which we could classify in 5 different classes for the game:</p>
            <ul>
                <li><label>small_intron is smaller than </label><input id="small_intron" type="number" value="5000">&nbsp;bps and normalized value is:<input id="small_intron-norm" type="number" value="1">units. With this type of intron we could say the avatar could easily jump.</li>
                <li><label>medium_intron is smaller than </label><input id="medium_intron" type="number" value="10000">&nbsp;bps and normalized value is:<input id="medium_intron-norm" type="number" value="5">units. With this type of intron we could say there are 1, 2 splicies (rocks) that the avatar need to jump on top to traverse the holw</li>
                <li><label>large_intron is smaller than </label><input id="large_intron" type="number" value="15000">&nbsp;bps and normalized value is:<input id="large_intron-norm" type="number" value="10">units. With this type of intron we could say that there is a spliceosome that moves</li>
                <li><label>big_intron is smaller than </label><input id="big_intron" type="number" value="20000">&nbsp;bps and normalized value is:<input id="big_intron-norm" type="number" value="20">units.  With this type of intron we could say that there is a spliceosome that moves and it is more complicated</li>
                <li><label>huge_intron is equal or bigger than </label><input id="huge_intron" type="number" value="20000">&nbsp;bps and normalized value is:<input id="huge_intron-norm" type="number"  value="50">units. With this type of intron there could be an SR protein before and the avatar would be propulsated (I think it is easy for the avatar)</li>
            </ul>
            <p>Exons (includes stop_codon, snp, start_codon) are relatively smallers (average is 500bps) compared to the introns (10000bps) and a continous normalisation approach was adopted:</p>
            <ul>
                <li><label>Division factor </label><input id="division-factor-exon" type="number" value="20">&nbsp; The length of the exons will be divided by this factor</li>
                <li><label>Minimal unit: </label><input style="font-size: 16px" id="minimal-unit" type="number" value="1">&nbsp; The exons will have in minimum this unit</li>
                <li><label>Maximal unit: </label><input style="font-size: 16px" id="maximal-unit" type="number" value="100">&nbsp; The exons will have in maximum this unit</li>
            </ul>
            <button onclick="loadProtein()" style="margin: 5px; font-size: 15px">Click here to normalize</button>
            <br>
        </div>
    </div>
    <div style="margin: 10px;">
        Based on your parameters: 
        Minimal units for a gene are: <span id="minUnits" style="color:green">computing...</span>
        Maximal units for a gene are: <span id="maxUnits" style="color:green">computing...</span>
        <br>Exons count with minimal unit: <span id="minExonsCount" style="color:green">computing...</span>
        Exons count with maximal unit: <span id="maxExonsCount" style="color:green">computing...</span> (minimize this number)
        <br>Introns types distribution: <span id="intronsDist" style="color:green">computing...</span>
    </div>
    <div style="margin: 10px;">
        <h3>Notes</h3>
        <ul>
            <li>Up / Down / Obstacles are not defined in this JSON, as agreed we leave that up to you to make sure the game is fun.</li>
            <li>Intron are much bigger than exons in reality (if possible in the drawing of introns could be a wrapped RNA like in the picture to seem bigger and maybe thiner, to discuss...)</li>
            <a target="_blank" href="data/assets/mockup.png"><img height="50px" src="data/assets/mockup.png"/></a>
            <li>There is not a hole below the start, snp stop codon (it is a limitation of the library in use for the schema). It should be a walkable exon with different color: from exon to utr (untranslated region)</li>
            <li>For SIB: Double Check aa length (does not seem correct for GJB2 counts 234aa but Ensembl counts 226aa)</li>
        </ul>
    </div>
    <div style="margin: 10px;">
        <br><button data-toggle='collapse' data-target="#full-json">Click to view and copy FULL JSON</button>
        <div style="font-size:8px" id="full-json" class='collapse'></div>
    </div>
    <div id="container" style="margin: 10px;">
    <h3>Genes and corresponding JSON</h3>
        <p>Genes are sorted by the length of the translated protein
        <br>
        Genes distribution by chromosome: <span id="genesDist" style="color:green">computing...</span>
        </p>
        <div id="genes"></div>
    </div>
</body>


<script>

    var maximalExon, minimalExon = 0;

    loadProtein()

    function getBaseLog(x, y) {
        return Math.log(y) / Math.log(x);
    }
    
    function normalizeLength(levelData){
        
        if (levelData.type === "intron") {
            var normalizedIntron = getNormalizedLengthForIntrons(levelData.realLength);
            levelData.class = normalizedIntron.class
            levelData.normalizedLength = normalizedIntron.normalizedLength
        } else {
            var normalized = getNormalizedLengthForNonIntron(levelData.realLength);
            levelData.normalizedLength = normalized.normalizedLength
            levelData.class = levelData.type
        }
        
    }

    function getNormalizedLengthForNonIntron(realLength) {

        var divisionFactor = parseInt($("#division-factor-exon").val())
        var minimalUnit = parseInt($("#minimal-unit").val())
        var maximalUnit = parseInt($("#maximal-unit").val())

        //Based on histogram in R
        //hist(log(exons$V6), breaks=5, col="lightgreen")
        var continousNormalization = Math.round(realLength / divisionFactor);
        if (continousNormalization > maximalUnit) {
            continousNormalization = maximalUnit;
            maximalExon++;
        }
        if (continousNormalization < minimalUnit) {
            continousNormalization = minimalUnit;
            minimalExon++;
        }
        

        return { normalizedLength: continousNormalization }

    };

    function getProteinLink (gene, source){

        var html, value, url, image;
        if(source === "chromosomeWalk") {
                image = "data/assets/chromosomeWalk.gif"
                var chr = gene.chromosome;
                if (gene.externalLinks && gene.externalLinks.chromosomeWalk) {
                    value = "Walk with the gene " + gene.geneName
                    url = gene.externalLinks.chromosomeWalk
                }else {
                    value = "Walk with the chromosome " + gene.chromosome
                    url = gene.externalLinks.chromosomeWalk
                }
        }
        else if(source === "pubmed" &&  gene.mutationInfo && gene.externalLinks.pubmed) {
                image = "data/assets/pubmed.png"
                value = "PubMed "
                url = gene.externalLinks.pubmed;
        }
        else if(source === "snpedia" &&  gene.mutationInfo && gene.mutationInfo.rsName) {
                image = "data/assets/snpedia.png"
                value = "Snpedia " + gene.mutationInfo.rsName
                url = gene.externalLinks.snpedia;

        }else if(source === "proteinSpotlight" &&  gene.externalLinks && gene.externalLinks.proteinSpotlight) {
            image = "data/assets/proteinSpotlight.png"
            value = "A story in Protein SpotLight";
            url = "http://web.expasy.org/spotlight/back_issues/" + gene.externalLinks.proteinSpotlight;
        }

        if(value) {
            return "<li><a target='_blank' href='"+ url + "'>"+
                "<img height='20px' src='" + image + "'</img>"+
                " "+ value  + " </a></li>";
        }else return ""

    }


    function getNormalizedLengthForIntrons(number) {

        //Based on histogram in R
        //hist(log(exons$V6), breaks=5, col="lightgreen")
        if (number < parseInt($("#small_intron").val()))
            return {
                class: "small_intron",
                normalizedLength: parseInt($("#small_intron-norm").val())
            } //simple jump --> most of them end up in this category
        if (number < parseInt($("#medium_intron").val()))
            return {
                class: "medium_intron",
                normalizedLength: parseInt($("#medium_intron-norm").val())
            } //mutiple jumps (kind of rocks) 
        if (number < parseInt($("#large_intron").val()))
            return {
                class: "large_intron",
                normalizedLength: parseInt($("#large_intron-norm").val())
            } //spliceosome
        if (number < parseInt($("#big_intron").val()))
            return {
                class: "big_intron",
                normalizedLength: parseInt($("#big_intron-norm").val())
            } //Complicated spliceosome
        return {
            class: "huge_intron",
            normalizedLength: parseInt($("#huge_intron-norm").val())
        } //SR protein (Catapult)
    };
    

    function renderGenes(genesInfo, genesData) {

        maximalExon = 0;
        minimalExon = 0;
        //Normalization and get max
        var maxNormalizedLength = 0;
        var minNormalizedLength = Number.MAX_SAFE_INTEGER;

        var totalClassCount = {};
        var totalChromosomeCount = {};

        genesData.forEach(function (gene) {
            var classCount = {};
            gene.totalNormalizedLength = 0
            gene.totalNormalizedWalkableLength = 0

            if(!totalChromosomeCount["Chr" + gene.chromosome]){
                totalChromosomeCount["Chr" + gene.chromosome] = 0;
            }
            totalChromosomeCount["Chr" + gene.chromosome] += 1;

            var snpType = "";
            var previousClass = "";
            var skipNextSum = false;
            gene.levelData.forEach(function (ld) {
                normalizeLength(ld);
                if(!classCount[ld.class]){
                    classCount[ld.class] = 0;
                } 
                if(!totalClassCount[ld.class]){
                    totalClassCount[ld.class] = 0;
                }

                if(!skipNextSum){
                    classCount[ld.class] += 1;
                    totalClassCount[ld.class] += 1;
                }

                gene.totalNormalizedLength += ld.normalizedLength
                if(ld.class.indexOf("intron") == -1 ){
                    gene.totalNormalizedWalkableLength += ld.normalizedLength
                }

                if(ld.class == "snp"){
                    if(previousClass.indexOf("_intron") != -1){
                        snpType = "intronic";
                    }else if(previousClass == "exon"){
                        snpType = "exonic";
                    }else if(previousClass == "utr"){
                        snpType = "utric";
                    }else {
                        console.log("WHAT IS THIS SNP???????", previousClass, gene.geneName)
                    }
                    skipNextSum = true;
                }else {
                    skipNextSum = false;
                }
                previousClass = ld.class;
            })

            gene.snpType = snpType;
            gene.classCount = classCount;
            //gene.classCount.exon = gene.classCount.exon - 1;
            gene.proteinLength = Math.ceil((gene.exonsRealLength + 1) / 3.0)
            var gi = genesInfo.filter(g => g.geneName === gene.geneName);
            Object.assign(gene, gi[0])

            maxNormalizedLength = Math.max(maxNormalizedLength, gene.totalNormalizedLength)
            minNormalizedLength = Math.min(minNormalizedLength, gene.totalNormalizedLength)

            if(gene.externalLinks && gene.externalLinks.uniprot){
                gene.externalLinks.nextprot = "https://www.nextprot.org/entry/NX_" + gene.externalLinks.uniprot
                gene.externalLinks.uniprot = "https://www.uniprot.org/uniprot/" + gene.externalLinks.uniprot
            }

            if(gene.externalLinks && gene.externalLinks.pubmed){
                gene.externalLinks.pubmed = "https://www.ncbi.nlm.nih.gov/pubmed/" + gene.externalLinks.pubmed
            }

            if(gene.externalLinks &&  gene.mutationInfo && gene.mutationInfo.rsName){
                gene.externalLinks.snpedia = "https://www.snpedia.com/index.php/" + gene.mutationInfo.rsName;
            }

            if(gene.externalLinks &&  gene.mutationInfoAlternative && gene.mutationInfoAlternative.rsName){
                gene.externalLinks.snpediaAlternative = "https://www.snpedia.com/index.php/" + gene.mutationInfoAlternative.rsName;
            }
            
            gene.externalLinks.transcriptLink = "http://www.ensembl.org/Homo_sapiens/transview?db=core;transcript=" + gene.transcript

        })
        
        genesSorted = genesData.sort(function(a,b){
            if (a.exonsRealLength < b.exonsRealLength) {
                return -1;
            }else if (b.exonsRealLength < a.exonsRealLength) {
                return 1;
            }
            return 0;
        });

        $("#maxUnits").text(maxNormalizedLength)
        $("#minUnits").text(minNormalizedLength)
        $("#genesDist").text(JSON.stringify(totalChromosomeCount));
        $("#maxExonsCount").text(maximalExon)
        $("#minExonsCount").text(minimalExon)
        $("#intronsDist").text("");
        $("#intronsDist").append("small_introns:" + totalClassCount["small_intron"] + " ")
        $("#intronsDist").append("medium_introns:" + totalClassCount["medium_intron"] + " ")
        $("#intronsDist").append("big_introns:" + totalClassCount["big_intron"] + " ")
        $("#intronsDist").append("large_introns:" + totalClassCount["large_intron"] + " ")
        $("#intronsDist").append("huge_introns:" + totalClassCount["huge_intron"] + " ")

        var g = 1;
        genesSorted.forEach(function (gene) {


            var sv = ""; 
            if(gene.superVariant) {
                sv = "<span style='background-color:orange'>super variant</span>";
            }


            function getMutationInfoHTML (mutationInfo){
                var tx = "";
                tx += "<br><br><b>Mutation properties (can be represented with symbols) on a little square:</b><ul>";
                tx += "<li>RS name: " + ((mutationInfo && mutationInfo.rsName) ? mutationInfo.rsName : "???")  + "</li>";
                tx += "<li>H/F applicable:";
                var sexNotSet = true;
                if(mutationInfo && mutationInfo.sex && mutationInfo.sex.indexOf("H") != -1) { tx += "&nbsp;&#9794;"; sexNotSet = false; }
                if(mutationInfo && mutationInfo.sex && mutationInfo.sex.indexOf("F") != -1) {  tx += "&nbsp;&#9792;"; sexNotSet = false; }
                if(sexNotSet){ html += "???"; }
                tx += "</li>";
                tx += "<li>Population Background:&nbsp;" + ((mutationInfo && mutationInfo.populationBackground) ? mutationInfo.populationBackground : "???")   + "</li>";
                tx += "<li>Alleles:</b> " + ((mutationInfo && mutationInfo.alleles) ? mutationInfo.alleles : "???")  + " </li>";
                tx += "<li>Type:&nbsp;" + ((mutationInfo && mutationInfo.type) ? mutationInfo.type : "???")   + "</li>";
                tx += "<li>Nucleotide change (from -> to) :&nbsp;" + ((mutationInfo && mutationInfo.nucleotideChange) ? (mutationInfo.nucleotideChange[0] + "->" + mutationInfo.nucleotideChange[1]) : "???")   + "</li>";
                tx += "</ul>"

                return tx;
    
            }
            html = ""


            html += "<div class='geneContainer'>" +

                     "<img class='pull-right' height='50px' src='data/assets/chromosomes/pair/chromosome_" + gene.chromosome + ".png'></img>" +  
                    "<img class='pull-right' height='50px' src='data/assets/chromosomes/single/chromosome" + gene.chromosome + ".png'></img>" +
                    "<h3>" + g++ + " - Gene name: " + gene.geneName  + " " + "&nbsp;" + sv;

            if(gene.debugInfo) {
                html += "&nbsp;<img height='30px' src='" + gene.debugInfo.img + "'></img>";
            }
            html +=  "</h3>"



            var codingPercentage = Math.round(( gene.exonsRealLength  / gene.geneRealLength) * 100);

            html += "<h4>Transcript: <a target='_blank' href='" + gene.externalLinks.transcriptLink + "'>" + gene.transcript + "</a></h4>";

            html +=  "<b>Chromosome</b>: " + gene.chromosome + "<br><b>Position in chromosome</b>: " + (gene.relativePositionPercentage * 100) + "%" + 
            "<a target='_blank' href='https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr" + gene.chromosome +  "%3A" + (gene.transcriptStartPosition - 10000) + "-" + (gene.transcriptEndPosition + 10000) + "'> localize it on UCSC Genome Browser</a>" +
            "<br> <b> Computed gene length (transcript end - start position) </b>: " +  (gene.transcriptStartPosition ? gene.transcriptStartPosition + "-" + gene.transcriptEndPosition + " => " + (gene.transcriptEndPosition -  gene.transcriptStartPosition) + "bp" : "????"); 
                    if((gene.transcriptEndPosition -  gene.transcriptStartPosition) != gene.geneRealLengh) {
                       html += " seems to be different than geneRealLength: " + gene.geneRealLength + " <span style='color:red'> TO CHECK.... </span>"
                    }
            html += "<br><b>Gene real length (summing blocks) :</b>" + gene.geneRealLength + " bp (base pairs can be A,C,G,T)  ";
            html += "<br><b>Coding Direction: </b>" + gene.transcriptDirection;

            html += "<br><br>THIS CAN BE CHANGED DEPENDING ON THE NORMALIZATION PARAMETERS:<br><b>Normalized Gene Length: </b>"+ gene.totalNormalizedLength + " units (walkable: " + gene.totalNormalizedWalkableLength + " units -> after removing introns), composed by <b>exons:</b> " + gene.classCount.exon + 
                    "&nbsp;&nbsp;<b>small_introns:</b> " + (gene.classCount["small_intron"] ? gene.classCount["small_intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>medium_introns:</b> " + (gene.classCount["medium_intron"] ? gene.classCount["medium_intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>large_introns:</b> " + (gene.classCount["large_intron"] ? gene.classCount["large_intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>big_introns:</b> " + (gene.classCount["big_intron"] ? gene.classCount["big_intron"] : "0")  + 
                    "&nbsp;&nbsp;<b>huge_introns:</b> " + (gene.classCount["huge_intron"] ? gene.classCount["huge_intron"] : "0")  + 
                    "<br><b>Sentence in the beginning of the game with the animation (TBC):</b><br><i> Only " + codingPercentage + "% of this gene encodes a protein: Those coding blocks, are called exons! The rest of the gene, is composed by introns and untranslated regions (UTR). <br>Once finished you will revealed the secret of this gene and associated protein! But to find out its secret, you will need first to avoid introns by jumping on top of them and collect the variant, start and stop codons! Good Luck!</i>";

            html += "<br><br>JSON:"
            html += "<br><span style='font-size:4px'>" + JSON.stringify(gene, null, 2) + "</span>";
            html += "<br><button style='font-size: 10px' data-toggle='collapse' data-target='#json" + gene.id + "'>Click to view JSON in pretty format</button>";
            html += "<div id='json" + gene.id + "' class='collapse'><pre><code>" + JSON.stringify(gene, null, 2) + "</code></pre></div>";
  
            html += "<div id=\'" + gene.id + "\'></div>";
            html += "Once spliced (introns are cut) the mRNA gets out of the nucleus and is translated by the ribosome (animation) which will render a protein of the following length:"
            html += "<br><b>Protein length:</b> (" +gene.exonsRealLength  +"bp/3)=" + Math.ceil((gene.exonsRealLength + 1) / 3.0)  + "aa (For every 3bp - base pairs the ribosome produces 1aa - amino acid).... According to <a target='_blank' href='http://knowpapa.com/wp-content/uploads/2013/10/6.png'>this table</a>."
            html += "<br><br><b>Description</b>:<br>" + gene.description ;
            if(gene.descriptionImg){ html += "<br><img width='250px' src='" + gene.descriptionImg + "'></img>"; }
            
            
            html += getMutationInfoHTML(gene.mutationInfo)
            if(gene.mutationInfoAlternative) {
                html += "<b style='color:red'>Alternative mutation</b>";
                html += getMutationInfoHTML(gene.mutationInfoAlternative)
            }
            html += "<br><b>STOP codon bases (can be UGA, UAA, UAG): </b>" + gene.stopCodon;
            html += "<br><br><b>Chromosome Walk</b>:<br>";
            html += ((getProteinLink(gene, "chromosomeWalk") !=="") ? getProteinLink(gene, "chromosomeWalk") : "???")

            html += "<br><b>Protein Resources Databases (links to external SIB resources):</b><br>" 
                 + getProteinLink(gene, "uniprot")  
                 + getProteinLink(gene, "nextprot") 
                 + getProteinLink(gene, "proteinSpotlight");

            html += "<br><b>Publication Studies</b>:<br>";
            html += ((getProteinLink(gene, "pubmed") !=="") ? getProteinLink(gene, "pubmed") : "???")
            html += ((getProteinLink(gene, "snpedia") !=="") ? getProteinLink(gene, "snpedia") : "???")
            
            html += "<br><b>Links to other resources (like swissdrug)</b><br>"

            if(gene.debugInfo) {
                html += "<br><b>Debug Info Image</b><br>";
                html += "Image to help understanding the impact of this mutation (NOT TO BE INCLUDED IN THE GAME): <br> <img height='50px' src='" + gene.debugInfo.img + "'></img>";
                if(gene.debugInfo.superVariantDescription){
                    html += ("<br>Super Variant Description: " + gene.debugInfo.superVariantDescription)
                }
            }

            html += "</div>";
            html += "<hr>";

            $("#genes").append(html);

            //Create a new Feature Viewer and add some rendering options
            var ft = new FeatureViewer(maxNormalizedLength, "#" + gene.id, {
                showAxis: true,
                showSequence: true,
                brushActive: true,
                toolbar: true,
                bubbleHelp: true,
                zoomMax: 10
            });

            var exonData = [];
            var intronData = [];
            var starsData = [];

            var cursor = 0;
            gene.levelData.forEach(function (ld) {
                
                var desc = ld.class ? ld.class : "" + "\n normalizedLength:" + ld.normalizedLength + "\n realLength:" + ld.realLength;
                var normalizedLength = ld.normalizedLength;
                if (ld.type === "exon" || ld.type === "utr") {
                    exonData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": (ld.type == "exon") ? "green" : "lightgreen"
                    })
                } else if (ld.type === "intron") {
                    intronData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": "blue"
                    })
                } else if (ld.type === "start_codon" || ld.type === "stop_codon" || ld.type === "snp") {
                    starsData.push({
                        "x": cursor,
                        "y": cursor + normalizedLength,
                        "description": desc,
                        "color": (ld.type == "snp") ? "red" : "orange",
                        "description": ld.type
                    })

                }
                cursor += normalizedLength
            })

            //Add some features
            ft.addFeature({
                data: starsData,
                name: "stars",
                type: "rect"
            });

            //Add some features
            ft.addFeature({
                data: exonData,
                name: "exons",
                type: "rect"
            });

            //Add some features
            ft.addFeature({
                data: intronData,
                name: "introns",
                type: "rect"
            });

        })

        $("#full-json").html("<pre><code>" +  JSON.stringify(genesSorted) + "</code></pre>")

    }

    function loadProtein() {
        $("#maxUnits").text("computing...")
        $("#minUnits").text("computing...")
        $("#genesDist").text("computing...")
        $("#intronsDist").text("computing...")
        $("#maxExonsCount").text("computing...")
        $("#minExonsCount").text("computing...")
        
        $("#genes").empty();
        $.getJSON('data/genes.json', function (genesInfo) {
            $.getJSON('data/data.json', function (genesData) {
                renderGenes(genesInfo, genesData);
            })
        });
    }
</script>
